app/hooks/header.ts

```diff
@@ -28,93 +29,95 @@ export const useDefaultHeaderHeight = () => {
     const insets = useSafeAreaInsets();
     const animatedRef = useAnimatedRef<Animated.ScrollView>();
-    const {largeHeight, defaultHeight} = useHeaderHeight(true, hasSubtitle, hasSearch);
+    const {largeHeight, defaultHeight} = useHeaderHeight();
     const scrollValue = useSharedValue(0);
+    const autoScroll = useSharedValue(false);
+    const snapping = useSharedValue(false);
+
+    const headerHeight = useDerivedValue(() => {
+        const minHeight = defaultHeight + insets.top;
+        const value = -(scrollValue?.value || 0);
+        const header = (isLargeTitle ? largeHeight : defaultHeight);
+        const height = header + value + insets.top;
+        if (height > header + (insets.top * 2)) {
+            return Math.min(height, largeHeight + insets.top + 80);
+        }
+        return Math.max(height, minHeight);
+    });
 
     function snapIfNeeded(dir: string, offset: number) {
         'worklet';
-        if (dir === 'up' && offset < defaultHeight) {
-            const diffHeight = largeHeight - defaultHeight;
-            let position = 0;
-            if (Platform.OS === 'ios') {
-                position = (diffHeight - (hasSearch ? -HEADER_SEARCH_BOTTOM_MARGIN : insets.top));
-            } else {
-                position = hasSearch ? largeHeight + HEADER_SEARCH_BOTTOM_MARGIN : diffHeight;
-            }
-            scrollTo(animatedRef, 0, position, true);
-        } else if (dir === 'down') {
-            const inset = largeHeight + (hasSearch ? HEADER_SEARCH_BOTTOM_MARGIN : 0);
-            if (offset < inset) {
-                scrollTo(animatedRef, 0, -insets.top, true);
+        if (onSnap && !snapping.value) {
+            snapping.value = true;
+            if (dir === 'down' && offset < largeHeight) {
+                runOnJS(onSnap)(0);
+            } else if (dir === 'up' && offset < (defaultHeight + insets.top)) {
+                runOnJS(onSnap)((largeHeight - defaultHeight));
             }
+            snapping.value = false;
         }
     }
 
     const onScroll = useAnimatedScrollHandler({
         onBeginDrag: (e: NativeScrollEvent, ctx: HeaderScrollContext) => {
             ctx.start = e.contentOffset.y;
+            ctx.dragging = true;
         },
-        onScroll: (e) => {
-            scrollValue.value = e.contentOffset.y;
+        onScroll: (e, ctx) => {
+            if (ctx.dragging || autoScroll.value) {
+                scrollValue.value = e.contentOffset.y;
+            } else {
+                scrollValue.value = 0;
+                scrollTo(animatedRef, 0, 0, false);
+            }
         },
         onEndDrag: (e, ctx) => {
-            if (ctx.start !== undefined && Platform.OS === 'ios') {
+            if (ctx.start !== undefined) {
                 const dir = e.contentOffset.y < ctx.start ? 'down' : 'up';
                 const offset = Math.abs(e.contentOffset.y);
-                ctx.start = undefined;
                 snapIfNeeded(dir, offset);
             }
```

Comment:
Shouldn't we do `ctx.draggint = false` here too?

Reply:
Not really, we want to set this onMomentumEnd, yes is true that we are not dragging anymore, but setting it to false here will cause problems with onScroll